clippy_lints/src/implicit_saturating_sub.rs

```diff
@@ -0,0 +1,152 @@
+use crate::utils::{higher, in_macro, match_qpath, span_lint_and_sugg, SpanlessEq};
+use if_chain::if_chain;
+use rustc_ast::ast::LitKind;
+use rustc_errors::Applicability;
+use rustc_hir::{BinOpKind, Expr, ExprKind, QPath, StmtKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for implicit saturating subtraction.
+    ///
+    /// **Why is this bad?** Simplicity and readability. Instead we can easily use an inbuilt function.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// let end: u32 = 10;
+    /// let start: u32 = 5;
+    ///
+    /// let mut i: u32 = end - start;
+    ///
+    /// // Bad
+    /// if i != 0 {
+    ///     i -= 1;
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// let end: u32 = 10;
+    /// let start: u32 = 5;
+    ///
+    /// let mut i: u32 = end - start;
+    ///
+    /// // Good
+    /// i = i.saturating_sub(1);
+    /// ```
+    pub IMPLICIT_SATURATING_SUB,
+    pedantic,
+    "Perform saturating subtraction instead of implicitly checking lower bound of data type"
+}
+
+declare_lint_pass!(ImplicitSaturatingSub => [IMPLICIT_SATURATING_SUB]);
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for ImplicitSaturatingSub {
+    fn check_expr(&mut self, cx: &LateContext<'a, 'tcx>, expr: &'tcx Expr<'tcx>) {
+        if in_macro(expr.span) {
+            return;
+        }
+        if_chain! {
+            if let Some((ref cond, ref then, None)) = higher::if_block(&expr);
+
+            // Check if the conditional expression is a binary operation
+            if let ExprKind::Binary(ref cond_op, ref cond_left, ref cond_right) = cond.kind;
+
+            // Ensure that the binary operator is >, != and <
+            if BinOpKind::Ne == cond_op.node || BinOpKind::Gt == cond_op.node || BinOpKind::Lt == cond_op.node;
+
+            // Check if the true condition block has only one statement
+            if let ExprKind::Block(ref block, _) = then.kind;
+            if block.stmts.len() == 1;
+
+            // Check if assign operation is done
+            if let StmtKind::Semi(ref e) = block.stmts[0].kind;
+            if let ExprKind::AssignOp(ref op1, ref target, ref value) = e.kind;
+            if BinOpKind::Sub == op1.node;
+            if let ExprKind::Path(ref assign_path) = target.kind;
```

Comment:
One nice and fairly easy enhancement here would be to extract this in a function `subtracts_one`, which also checks for `ExprKind::Assign(.., expr)`, where the assigned `expr` is a `ExprKind::BinOp(Sub, ..)`. With this you have to match the left side of the substraction, so that it is the same as the assigned binding.